Neos.Neos/Classes/Domain/Service/WorkspaceService.php

```diff
@@ -0,0 +1,365 @@
+            )
+        );
+        $this->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, $classification, $ownerId);
+        return $workspaceName;
+    }
+
+    private function addWorkspaceMetadata(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, WorkspaceTitle $title, WorkspaceDescription $description, WorkspaceClassification $classification, UserId|null $ownerUserId): void
+    {
+        try {
+            $this->dbal->insert(self::TABLE_NAME_WORKSPACE_METADATA, [
+                'content_repository_id' => $contentRepositoryId->value,
+                'workspace_name' => $workspaceName->value,
+                'title' => $title->value,
+                'description' => $description->value,
+                'classification' => $classification->name,
+                'owner_user_id' => $ownerUserId?->value,
+            ]);
+        } catch (DbalException $e) {
+            throw new \RuntimeException(sprintf('Failed to add metadata for workspace "%s" (Content Repository "%s"): %s', $workspaceName->value, $contentRepositoryId->value, $e->getMessage()), 1727084068, $e);
+        }
+    }
+
+    private function findPrimaryWorkspaceNameForUser(ContentRepositoryId $contentRepositoryId, UserId $userId): ?WorkspaceName
+    {
+        $tableMetadata = self::TABLE_NAME_WORKSPACE_METADATA;
+        $query = <<<SQL
+            SELECT
+                workspace_name
+            FROM
+                {$tableMetadata}
+            WHERE
+                content_repository_id = :contentRepositoryId
+                AND classification = :personalWorkspaceClassification
+                AND owner_user_id = :userId
+        SQL;
+        $workspaceName = $this->dbal->fetchOne($query, [
+            'contentRepositoryId' => $contentRepositoryId->value,
+            'personalWorkspaceClassification' => WorkspaceClassification::PERSONAL->name,
+            'userId' => $userId->value,
+        ]);
+        return $workspaceName === false ? null : WorkspaceName::fromString($workspaceName);
+    }
+
+    private function loadWorkspaceRoleOfUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, User $user): ?WorkspaceRole
+    {
+        try {
+            $userRoles = array_keys($this->userService->getAllRoles($user));
+        } catch (NoSuchRoleException $e) {
+            throw new \RuntimeException(sprintf('Failed to determine roles for user "%s", check your package dependencies: %s', $user->getId()->value, $e->getMessage()), 1727084881, $e);
+        }
+        $tableRole = self::TABLE_NAME_WORKSPACE_ROLE;
+        $query = <<<SQL
+            SELECT
+                role
+            FROM
+                {$tableRole}
+            WHERE
+                content_repository_id = :contentRepositoryId
+                AND workspace_name = :workspaceName
+                AND (
+                    (subject_type = :userSubjectType AND subject = :userId)
```

Comment:
it's not outdated, I just removed the `OWNER` role and made that part of the metadata because that had some weird implications.
But it should still be possible to grant a single user specific permissions on a ws

Reply:
ahh i see i was just confused that there is no write api for it ... just that we query this ;)
Thanks for thinking two steps ahead! ❤️